docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,130 @@
+
+## Problem Description
+
+The current approach of syncing the spec from registries is as follows:
+ 1. Delete all the specs currently in the datastore
+ 2. Fetch new specs
+ 3. Add all the fetched specs to the datastore
+ 
+This leads to a bunch of bugs:
+ - [Bug 1583495](https://bugzilla.redhat.com/show_bug.cgi?id=1583495)
+ - [Bug 1577810](https://bugzilla.redhat.com/show_bug.cgi?id=1583495) 
+ - [Bug 1586345](https://bugzilla.redhat.com/show_bug.cgi?id=1586345)
+
+The underlying problem with all those bugs is that the broker is deleting specs even if it is provisioned. When the catalog tries to deprovision, bind, unbind with these dangling instances (instances with no specs) it will fail. Hence the broker needs a way to safely delete the specs.
+
+### Approaches to safely deleting specs
+
+The broker can take the following actions to safely delete the specs:
+#### Approach#1
+ 1. Fetch the specs from registry
+    1. If the fetch fails increment a failed connection counter. If the failed connection counter reaches a limit delete all the specs that do not have a service instance
+    2. If the fetch succeeds:
+        1. add new specs to the datastore
+        2. delete that specs that are removed from datastore and do not have a service instance.
+        3. mark the specs that are removed from datastore and has a service instance
+ 2. During deprovision, if the instance getting deprovisioned is the last one delete the spec from datastore
+ 3. Serve the catalog request with specs that are not marked
+ 
+This will solve the problem but the task of deleting the specs is spread across time (when fetching specs and when deprovisioning)
+
+#### Approach#2
+ 1. Fetch the specs from registry
+    1. If the fetch fails mark all the specs for deletion
+    2. If the fetch succeeds,
+       1. add new specs to the datastore
+       2. mark the specs that are removed from the registry.
+ 3. Run a background task to delete all the marked specs that does not have a service instance currently provisioned. The interval of running this task can be configurable
+ 4. Serve the catalog request with specs that are not marked
+ 
+This solves the problem but introduces a new parameter of configuration. 
+
+If in future, other conditions of deleting a spec crop up, just mark it for deletion and background task will take care of deleting it safely.
+
+
+Either approach will take care of the all objectives and bugs. 
+
+### Approaches to handle untimely provision request
+
+
+To to handle a case where provision is requested after the broker has started bootstrapping but before it has completed bootstrapping, the broker can take the following approaches:
+
+#### Approach#1
+The broker always sends an error code if it is busy provisioning. OSB [spec](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-2) does not have an appropriate error code for busy.
+
+#### Approach#2 
```

Comment:
I think that a combination of 2 and 3 is an option. in this scenario we start the provision with some soft of waiting mechanism, that will get notified when bootstrap finishes. While this is happening async provision last operation request's can be called just like they are now and will error if the spec is deleted.

Reply:
Agreed with the above comments. I don't think 1 is reasonable, 2 shouldn't block. It would be semantically correct to respond with a 202 saying "okay, we've accepted your request and we're working on it". The fact that we're waiting on the bootstrap to complete is a broker implementation detail; it has nothing to do with the OSB spec.

Once the bootstrap completes, the job waiting should then investigate the current state of the specs. If the requested instance refers to a spec that has been deleted, we can mark the job failed. If the spec still exists, proceed as normal.

Reply:
I like approach 3 the best and agree with the above comments.